BaseHtml fix renderSelectOptions for Enum

ionutwho opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  BaseHtml fix renderSelectOptions for Enum | <!-- comma-separated list of tickets # fixed by the PR, if any -->

ionutwho committed:
Update BaseHtml.php

ionutwho commented on framework/helpers/BaseHtml.php:
```diff
@@ -1935,7 +1935,7 @@ public static function renderSelectOptions($selection, $items, &$tagOptions = []
                             $selected = !strcmp((string)$key, (string)$selection);
                         } else {
-                            $selected = $selection == $key;
+                            $selected = is_object($selection) ? $selection->value == $key : $selection == $key;
```
Do you mean to check if its a Instance of Enum?

ionutwho replied:
added extra check for object: check if it has value property

ionutwho committed:
Update BaseHtml.php

added extra check: check the object has value property

bizley commented on framework/helpers/BaseHtml.php:
```diff
@@ -1935,7 +1935,8 @@ public static function renderSelectOptions($selection, $items, &$tagOptions = []
                             $selected = !strcmp((string)$key, (string)$selection);
                         } else {
-                            $selected = is_object($selection) ? $selection->value == $key : $selection == $key;
+                            $selected = is_object($selection) && property_exists($selection, 'value') ?
```
instanceof would be better

cristiangirlea replied:
@bizley  Can you check the latest changes

cristiangirlea committed:
Update BaseHtml.php

cristiangirlea reviewed (commented):
Pull Request: Improve Selection Matching Logic in renderSelectOptions
Description:
This PR refines the logic for determining if an option is selected in the renderSelectOptions function. The changes ensure clearer and more concise comparisons, improving code readability.

Changes:
Introduced an additional condition to handle cases where $selection is an object with a value property.
Simplified the comparison logic for when $strict is false.

cristiangirlea committed:
Update HtmlTest.php

Pull Request Description
Changes:
Enhanced the selection check in Html::renderSelectOptions:
Added a condition to handle cases where $selection is an object that possesses a value property.
Rationale:
The primary goal of these changes is to provide improved flexibility in the types of data that can be passed as $selection and to streamline the comparison logic for better maintainability and clarity.

Test Additions:
Introduced a new test, to verify the behavior when $selection is an object with a value property.

cristiangirlea reviewed (changes requested):
Pull Request: Improve Selection Matching Logic in renderSelectOptions
Description:
This PR refines the logic for determining if an option is selected in the renderSelectOptions function. The changes ensure clearer and more concise comparisons, improving code readability.

Changes:
Introduced an additional condition to handle cases where $selection is an object with a value property.
Simplified the comparison logic for when $strict is false.

cristiangirlea committed:
Merge branch 'master' into changes_in_base_html

ionutwho commented:
> I also think that using `instanceof UnitEnum` would be better - it should be safer than changing behavior for all objects.
> 
> And IMO it should be done at the beginning of `$selection` processing:
> 
> ```
> if ($selection !== null) {
>     if ($selection instanceof \UnitEnum) {
>         $selection = $selection->value;
>     }
>     ...
> ```

Well i tried your fix but its not working because we are using this: https://github.com/spatie/enum 
Its not really an Enum, actually its a Class extending an Enum class which doesnt implement UnitEnum...
`abstract class Enum extends \Spatie\Enum\Enum`
`abstract class Enum implements JsonSerializable
`

bizley commented:
I don't think we should support any object here, this should be out of the scope of the framework and it should depend on the developers to extend framework functionality in their own app to support any custom solution like in your case. Still, we would accept real enums here.

Webkadabra commented:
are you submitting fixes for some of your internal code to the public? Why are you not checking the type of `value` property then? 
Another useless PR. @samdark I hope we're not thinking of adapting this

terabytesoftw commented:
Is this PR ready?

bizley commented:
I'm closing it then. Still, PR introducing support for enums is welcome.
